cpp/src/arrow/util/logging.h

```diff
@@ -86,56 +67,40 @@ enum class ArrowLogLevel : int {
 #define ARROW_CHECK_GE(val1, val2) ARROW_CHECK((val1) >= (val2))
 #define ARROW_CHECK_GT(val1, val2) ARROW_CHECK((val1) > (val2))
 
-#ifdef NDEBUG
-#define ARROW_DFATAL ::arrow::util::ArrowLogLevel::ARROW_WARNING
+#define ARROW_LOG_FAILED_STATUS(status, ...)                                        \
+  ARROW_LOG(FATAL) << " Operation failed: " << ARROW_STRINGIFY(__VA_ARGS__) << "\n" \
+                   << " Bad status: " << _s.ToString()
+
+// If the status is bad, CHECK immediately, appending the status to the logged
+// message.
+#define ARROW_CHECK_OK(...)                                                    \
+  for (::arrow::Status _s = ::arrow::internal::GenericToStatus((__VA_ARGS__)); \
```

Comment:
Is there a particular reason for using a for loop?

Reply:
Yes, it allows us to confine `_s` to the scope of the loop body. The loop body is a single statement with no braces, allowing `*CHECK_OK()` to accept a stream.

Reply:
Ok, but I mean compared to a regular `while` or `do...while`?

Reply:
The last tokens of a `do... while` loop are the loop condition, which precludes providing a stream for extra error messages.

`for` supports a scoped decl, which `while` doesn't. If we wanted to avoid `for`, we could introduce a variable as in `ARROW_ASSIGN_OR_RAISE`. After some macro expansion that'd look like:

```c++
Status _s_56 = Thing();
if (ARROW_PREDICT_FALSE(!_s_56.ok()))
ARROW_LOG(FATAL) << _s.message() << "Extra messages";
```